src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -173,15 +179,159 @@ public async Task RefreshAsync(CancellationToken cancellationToken)
+                        .ConfigureAwait(false);
+
+                    if (!refreshAll)
+                    {
+                        applicationSettings = new Dictionary<string, ConfigurationSetting>(_applicationSettings, StringComparer.OrdinalIgnoreCase);
+
+                        foreach (KeyValueChange change in keyValueChanges.Values.Concat(changedKeyValuesCollection))
+                        {
+                            if (change.ChangeType == KeyValueChangeType.Deleted)
+                            {
+                                applicationSettings.Remove(change.Key);
+                            }
+                            else if (change.ChangeType == KeyValueChangeType.Modified)
+                            {
+                                applicationSettings[change.Key] = change.Current;
+                            }
+
+                            // Invalidate the cached Key Vault secret (if any) for this ConfigurationSetting
+                            foreach (IKeyValueAdapter adapter in _options.Adapters)
+                            {
+                                adapter.InvalidateCache(change.Current);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // Invalidate all the cached KeyVault secrets
+                        foreach (IKeyValueAdapter adapter in _options.Adapters)
+                        {
+                            adapter.InvalidateCache();
+                        }
+                    }
+
+                    if (_options.Adapters.Any(adapter => adapter.NeedsRefresh()) || changedKeyValuesCollection?.Any() == true || keyValueChanges.Any())
+                    {
+                        SetData(await PrepareData(applicationSettings, cancellationToken).ConfigureAwait(false));
+
+                        _applicationSettings = applicationSettings;
+
+                        foreach (KeyValuePair<KeyValueWatcher, KeyValueChange> kvp in keyValueChanges)
+                        {
+                            KeyValueChange keyValueChange = kvp.Value;
+                            KeyValueWatcher changeWatcher = kvp.Key;
+                            KeyValueIdentifier kvIdentifier = new KeyValueIdentifier(changeWatcher.Key, changeWatcher.Label);
+
+                            if (keyValueChange.ChangeType == KeyValueChangeType.Modified)
+                            {
+                                _watchedSettings[kvIdentifier] = keyValueChange.Current;
+                            }
+                            else if (keyValueChange.ChangeType == KeyValueChangeType.Deleted)
+                            {
+                                _watchedSettings.Remove(kvIdentifier);
+                            }
+
+                            UpdateCacheExpirationTime(kvp.Key);
+                        }
+                    }
                 }
                 finally

```

Comment:
How do we guarantee we also extend the cache expiration even on the failure case?

Reply:
Updating cache expiration on a failure case was a mechanism we used to use as a backoff mechanism. Now that we are utilizing backoff mechanism from the `ConfigurationClientManager`, we are not updating the cache expiration time on failure.

Reply:
When the call to AppConfig succeeds, but a call to KeyVault fails, we throw an exception and that exception to bubbled up to the caller and we don't update the `cacheExpirationTime` for the key value, so when the next `RefreshAsync` method is called, we try to get the key-value and its related key-vault secret again.

Reply:
If refreshAll is true, you update watchers before calling PrepareData (where keyvault is updated). If refreshAll is false, you update watchers after calling PrepareData.

I think we should always update watchers before calling PrepareData because repeating appconfig calls (by not updating watchers) won't help anything for keyvault calls. As long as adapter.NeedsRefresh is true, we will attempt to update keyvault again the next time RefreshAsync is called anyway.

Reply:
Thoughts on this (in case this is missed)?

Reply:
Yes, I am updating it.

Reply:
Done!